tests/lac/tensor_product_matrix_07.cc

```diff
@@ -0,0 +1,207 @@
+            (fe_values.shape_grad(i, q_index) *
+             fe_values.shape_grad(j, q_index) * fe_values.JxW(q_index));
+        }
+
+
+
+  return {mass_matrix_reference, derivative_matrix_reference};
+}
+
+template <typename Number>
+void
+print(const FullMatrix<Number> &matrix)
+{
+  matrix.print_formatted(deallog.get_file_stream(), 2, false, 7);
+  deallog << std::endl << std::endl;
+}
+
+
+void
+do_test(const bool zero_out_constraints)
+{
+  using Number                 = double;
+  const unsigned int dim       = 2;
+  const unsigned int n_rows_1d = 4;
+
+  // compute 2D stiffness matrix
+  const auto refernce_matrices_2D =
+    compute_reference_matrices<dim, Number>(n_rows_1d - 1);
+  auto K_2D = std::get<1>(refernce_matrices_2D);
+
+  // ... and apply DBC
+  for (unsigned int j = 0; j < Utilities::pow(n_rows_1d, dim); ++j)
+    for (unsigned int i = 0; i < Utilities::pow(n_rows_1d, dim - 1); ++i)
+      {
+        K_2D[i][j] = 0.0;
+        K_2D[j][i] = 0.0;
+      }
+
+  if (zero_out_constraints == false)
+    for (unsigned int i = 0; i < Utilities::pow(n_rows_1d, dim - 1); ++i)
+      K_2D[i][i] = 1.0;
+
+  // ... print
+  print(K_2D);
+
+  if (zero_out_constraints)
+    for (unsigned int i = 0; i < Utilities::pow(n_rows_1d, dim - 1); ++i)
+      K_2D[i][i] = 1.0;
+
+  K_2D.gauss_jordan();
+
+  if (zero_out_constraints)
+    for (unsigned int i = 0; i < Utilities::pow(n_rows_1d, dim - 1); ++i)
+      K_2D[i][i] = 0.0;
+
+  // ... print
+  print(K_2D);
+
+  // compute 1D stiffness and mass matrix
+  const auto refernce_matrices_1D =
+    compute_reference_matrices<1, Number>(n_rows_1d - 1);
+
+  // ... setup FDM
+  std::array<Table<2, Number>, dim> mass_matrix;
+  std::array<Table<2, Number>, dim> derivative_matrix;
+
+  for (unsigned int d = 0; d < dim; ++d)
+    {
+      mass_matrix[d]       = std::get<0>(refernce_matrices_1D);
+      derivative_matrix[d] = std::get<1>(refernce_matrices_1D);
+
+      // apply constraints
+      if ((d + 1) == dim)
+        {
+          for (unsigned int i = 0; i < n_rows_1d; ++i)
+            {
+              mass_matrix[d][i][0]       = 0.0;
+              mass_matrix[d][0][i]       = 0.0;
+              derivative_matrix[d][i][0] = 0.0;
+              derivative_matrix[d][0][i] = 0.0;
+            }
+
+          if (zero_out_constraints == false)
+            {
+              mass_matrix[d][0][0]       = 1.0;
+              derivative_matrix[d][0][0] = 1.0;
+            }
+        }
+    }
+
+  // ... print matrix
+  TensorProductMatrixSymmetricSum<dim, Number, -1> fdm;
+  fdm.reinit(mass_matrix, derivative_matrix);
+
+  FullMatrix<Number> matrix(fdm.m(), fdm.m());
+
+  for (unsigned int i = 0; i < fdm.m(); ++i)
+    {
+      Vector<Number> dst(fdm.m());
+      Vector<Number> src(fdm.n());
+
+      src[i] = 1.0;
+
+      fdm.vmult(make_array_view(dst), make_array_view(src));
+
+      for (unsigned int j = 0; j < fdm.m(); ++j)
+        matrix[j][i] = dst[j];
+    }
+
+  print(matrix);
```

Comment:
I think printing some text describing what the respective matrix in the output stream refers to would be helpful.